Collect and convert results into jUnit format

cevich opened the pull request:
Also, include in a comment, the old (kind of working) Jenkins
result-verification script.  When translated into tasks,
the intention is running this will produce an exit status
that matches both result existance and PASS status.

Signed-off-by: Chris Evich <cevich@redhat.com>

cevich commented:
Tossed this together super-quick, it's incomplete, decidedly not pretty, and it's untested.  

WIP: Push early, push often 😀

cevich commented:
--fixup the "not working" part - this will successfully collect results

cevich commented:
note-to-self: https://github.com/ansible/ansible/blob/stable-2.3/lib/ansible/plugins/callback/junit.py

cevich commented:
Ahh, this one seems much better designed and active:  https://github.com/kyrus/python-junit-xml.git

cevich commented:
All - annotating code...feedback welcome.

cevich commented on contrib/test/crio-integration-results.py:
```diff
@@ -0,0 +1,26 @@
+#    existing_xml = JunitXml(junit_output_filepath)
+#    for test_suite in new_xml.test_suite:
+#       existing_xml.add_testsuite(test_suite)
+#    exit(0)  -  close files / release lock
```
This will make more sense if you read the other diff's first...

cevich commented on contrib/test/crio-integration-results.yml:
```diff
@@ -0,0 +1,66 @@
+      tags:
+          - e2e
+      set_fact:
+        result_filepaths: "{{ result_filepaths + [crio_node_e2e_filepath] }}"
```
Tagging these tasks using the same ones from #927 to support running this under CI as separate jobs --or-- as a single-job that includes all hosts.  In other words, ``result_filepaths`` either ends up containing a single file path or two (or more if added later).

cevich commented on contrib/test/crio-integration-results.yml:
```diff
@@ -0,0 +1,66 @@
+          msg: "Cowardly refusing to overwrite {{ item }}"
+      when: item | exists
+      delegate_to: localhost
+      with_items: '{{ _dstfnbuff }}'
```
This will be fatal, no junit output will be produced, CI must be sensitive to exit-status of this playbook.

cevich commented on contrib/test/crio-integration-results.yml:
```diff
@@ -0,0 +1,67 @@
+
+    - name: Convert results into jUnit format
+      command: '{{ playbook_dir }}/crio-integration-results.py {{ result_filepaths | join(' ') }} results/results.xml'
+      delegate_to: localhost
```
JFM: Run this script locally for all playbook hosts (in parallel), the script does flocking on ``results.xml`` update.

cevich commented on contrib/test/requirements.txt:
```diff
@@ -52,3 +52,7 @@ virtualenv==15.1.0 --hash=sha256:39d88b533b422825d644087a21e78c45cf5af0ef7a99a1f
+
+future==0.16.0 --hash=sha256:e39ced1ab767b5936646cedba8bcce582398233d6a627067d4c6a454c90cfedb
+
+junitparser==1.0.0 --hash=sha256:5b0f0ffeef3548878b5ae2cac40b5b128ae18337e2a260a8265f5519b52c907c
```
future is needed by junitparser.   I  inspected the 1.0.0 code on github and didn't see anything nefarious.

cevich commented on contrib/test/venv-ansible-playbook.sh:
```diff
@@ -46,11 +46,14 @@ else
 
+# Create a directory to contain logs and test artifacts
+mkdir -p "$WORKSPACE/results"
+
```
This may need coordination with #927 - I'm not married to the location, but we should pick something and stick to it.

rhatdan commented:
@cevich Should we test this?

cevich commented:
Split the collection-aspect off into a separate PR: #945

This one will just focus on producing the canonical jUnit XML for consumption by ci-automation.

cevich commented:
...replies to @stevekuznetsov from other PR, moved here:

> Neither our Jenkins nor the Gubernator pages need extra information

Possibly, it depends on how the results are presented.  Unfortunately, the producer of the results can't always know where/how they will be consumed.  In most cases, it's very easy to "assume" the source, especially when it's not recorded in the data.

Since I'm parsing results anyway, it's just as easy to stick an extra XML attribute into the elements that *proves* where they came from.  Pinning the results to a specific hostname for example.

I would consider not having that extra detail as a very important near-term goal, if not a blocker.  

Results of questionable or unverifiable origin are as good as a soup-sandwich.  Esp. since this group already had a related run-in, that was very scary.

What is this Gubernator thing, such a sexy name, I simply must know more 😀 

> Is this separate from the jUnit pull request to BATS

I could easily be wrong about BATS.  By my read/investigation, it appears as an after-through addition, conditional on ``xunit == True``.  A leftover vestige from a bygone Jenkins era that's probably over-complicating what a simple script can do.  Effectively dead-code once this PR goes in.  Though if my understanding is wrong, it would be nice to know early 😀

stevekuznetsov commented:
> Possibly, it depends on how the results are presented. 

The three main ways engineers use these results are:
 - in legacy jobs, through the Jenkins UI
 - using Gubernator
 - through TestGrid

In all cases, there is a base assumption that the execution environment is not salient to the test failure and that test runs are independent. Therefore, seeing test X fail 10 times in a row indicates something wrong with the code under test (or framework executing it) not an issue with _e.g._ a Jenkins node.

>  Pinning the results to a specific hostname for example.

Tests are run on ephemeral VMs in EC2 so I'm not sure what this gains us, but sure, there's nothing wrong with putting more information into the file as long as that is not a blocker for collecting it.

> By my read/investigation, it appears as an after-through addition, conditional on xunit == True

I have no knowledge of this, but from the quick look I did, there is a PR/fork of BATS with jUnit support. Seems like _that_ is an easier place to integrate this sort of thing than parsing the output file from a BATS run if the PR is something that will be accepted. If it's dead and the project doesn't want it, your approach here sounds good.

cevich commented:
> execution environment is not salient to the test failure

Yep, that's exactly the issue that arises when tabulating multiple jUnits.  They silently become dependent on both the execution environment and run-independence (among other factors).  Here's an example.  Consider what happens when these two (somewhat fabricated) e2e jUnits are tabulated into a single "result" (I'm assuming worst-case, one-run, multiple hosts running the e2e tests):

Host A:
```XML
<?xml version="1.0" encoding="UTF-8"?>
  <testsuite tests="1" failures="0">
      <testcase name="[k8s.io]..." classname="Kubernetes e2e suite" time="0"></testcase>
</testsuite>
```

Host B:
```XML
<?xml version="1.0" encoding="UTF-8"?>
  <testsuite tests="1" failures="1">
      <testcase name="[k8s.io]..." classname="Kubernetes e2e suite" time="0">
             <failure message="test failure">Assertion failed</failure>
      </testcase>
</testsuite>
```

The tabulated result either shows the "[k8s.io]..." testcase passed or failed depending on who/what/how the the testing system handled the files.  


The fix is super-simple, add something unique (like a hostname or machine-id or whatever) to all the results:  Like this:

Host A:
```XML
<?xml version="1.0" encoding="UTF-8"?>
  <testsuite tests="1" failures="0">
      <testcase name="(Host A)[k8s.io]..." classname="Kubernetes e2e suite" time="0"></testcase>
</testsuite>
```

Host B:
```XML
<?xml version="1.0" encoding="UTF-8"?>
  <testsuite tests="1" failures="1">
      <testcase name="(Host B)[k8s.io]..." classname="Kubernetes e2e suite" time="0">
             <failure message="test failure">Assertion failed</failure>
      </testcase>
</testsuite>
```

Now the tabulated result shows that one host passed but the other failed.  Since the identifying information is tied to the result, debugging is made that much simpler.

> as long as that is not a blocker for collecting it.

Oh no, not for collection.  That's why I split up / re-purposed these PRs 😀

rhatdan commented:
I need this to keep my sanity.  Grogging through logs to find out if something failed is not fun.

stevekuznetsov commented:
@cevich strong NACK to changing the NAME of a test ... upstream tools like testgrid absolutely need to be able to consider a test run on host A and host B as the same test and look at the results together. Again, we run with ephemeral VMs. There is never a case where a specific host is broken or causes tests to fail. Te architecture does not allow for it. Upstream tooling to track test failures over time has a strong pre-requisite that test runs are independent. If we cannot deliver that with CI (we can), we should not be running it. I don't see an issue with putting new XML attributes on the tests as long as they fit the schema, but changing things like name fields is a bad idea.

cevich commented:
@stevekuznetsov I think that's okay, we can put the distinguishing detail somewhere else, like the testsuite name or 'classname'.  To be clear, the issue here isn't mixing up results from one run to the next, it's about not letting results from one host get masked by another from the same run.  So, confusing a RHEL 7.4 pass a 7.3 pass instead, or (god-forbid) a Fedora pass.  When the jUnits are combined, unless there's some distinguishing feature to each testcase or testsuite, they all get mixed up, leaving @rhatdan searching for hairs to pull 😀

My secondary motivation here is it gives us an interpretation-layer where we could (later) massage RHEL z-stream results.  For example, if in 7 years there's a bug that won't ever be fixed in 7.3.z. It's easier to mask the bogus failure at this layer, rather than try and fiddle with the tests - which work fine everywhere else.

cevich commented:
@stevekuznetsov I think that's okay, I can put the distinguishing detail somewhere else, like the ``testsuite`` name or classname or wherever.  This is what the gubinator(?) thing does:

https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/16408/test_pull_request_origin_extended_conformance_crio/15/

The issue I'm trying to solve here isn't mixing up results from one run to the next, it's about not letting results from one host get masked by another from the same run.  Let's sync-up and see what we can do here, maybe there's a have-cake-eat-cake option.  For example, could gubinator consume a different jUnit from Jenkins?  Maybe there's another jUnit key I can fuss with that gubinator ignores but xUnit does not.

stevekuznetsov commented:
None of our tools combine jUnit between jobs. A 7.3 job will only have data fr builds of that job, for instance. See the test grid.. for instance 

https://k8s-testgrid.appspot.com/kubernetes-presubmits#pull-kubernetes-e2e-gce-etcd3
https://k8s-testgrid.appspot.com/kubernetes-presubmits#pull-kubernetes-e2e-gce-bazel

Running the same tests, but in different contextx

TomSweeneyRedHat commented:
@cevich needs a rebase

cevich commented:
@stevekuznetsov Yes, context, that's what I want to preserve 😀 

I'm assuming gubinator made those?  (I still dunno what it is 😕)  So looking at the first one, how is a developer to know those results came from a Fedora host vs a RHEL host, and which release it was?

What about the idea of feeding Gubinator a different jUnit file from what Jenkins consumes?  Could that work?  Smells like an easy solution to me.

stevekuznetsov commented:
Developers in k8s are not interacting with Jenkins. The tests in question assume there is no difference between them running on Fedora, or RHEL, or whatever -- I believe those tests specifically run on a Debian-based distro.

Gubernator shows jUnit output for a specific build of a specific job. TestGrid aggregates results per job. If you wanted to test both Fedora and RHEL, you'd have separate jobs to do it.

cevich commented:
Thanks for the extra detail @stevekuznetsov so in the TestGrid, there's no association of the results to a platform or release, other than a job-name or number?  When viewing a grid for a build, how much digging (no. clicks) would a developer need to make to find out what distro/release it ran on?

The TestGrid is neat, and seems useful for giving a good overall picture of what's going on, I don't want to step on that at all.  The problem I'm trying to solve in this PR, is helping a CRI-O PR author to quickly find the source to any specific test result on a playbook run (possibly across more than one host).

IIRC, you/someone mentioned that the gubinator/grid only consumes jUnits of a particular name?
Is there some reason you can think of why we can't simply tabulate the results how we need them, into a different jUnit file from what gubinator consumes?  That would allow developers to mold the results as needed without affecting the testgrid.

i.e. The latest rebase/push (@rhatdan: code-complete but un-tested) would leave the following for archiving / further processing.

```
$WORKSPACE/artifacts/main.log                  # playbook log
$WORKSPACE/artifacts/canonical_results.xml     # junit_01.xml + testout.txt w/ hostnames & date/time
$WORKSPACE/artifacts/<hostname>/testout.txt    # raw integration testing output
$WORKSPACE/artifacts/<hostname>/junit_01.xml   # raw unmodified output
```

My preference would be to also add hostnames and date/time to the ``junit_01.xml`` but I can live without that as long as Jenkins is consuming (only) ``canonical_results.xml``.

cevich commented:
Note to me:  Waiting for #945 so I can get my hands on a real ``testout.txt`` and ``junit_01.xml`` to play with.

stevekuznetsov commented:
> When viewing a grid for a build, how much digging (no. clicks) would a developer need to make to find out what distro/release it ran on?

If it's an important part of the job, it will be in the job name usually. For many of these tests, this is not an important axis of variance.

> Is there some reason you can think of why we can't simply tabulate the results how we need them, into a different jUnit file from what gubinator consumes?

This approach sounds good to me.

cevich commented:
rebase, ``--autosquash`` and ``--force`` push.

cevich commented:
sigh...travis, travis, travis, what are you trying to say?

cevich commented:
Travis-translation (according to jhonce): "I suspect your using <1.8.1"

Since that build was suppose to be for "tip", and this PR has nothing to do with building stuff...that sounds bad...need to dig more 😖

cevich commented:
OIC the problem:  The integration build failed trying to find the e2e output (and vis versa).  

@stevekuznetsov when you get a sec. (it's not urgent).  I see the ``results.yml`` playbook is being called w/o the tags used on the main job, e.g.

```
ansible-playbook -vv --become -i localhost, --tags all --become-user root --connection local /go/src/github.com/kubernetes-incubator/cri-o/contrib/test/integration/results.yml
```

Is it a terrible pain to have that use the same tags passed to ``main.yml``?  If so, then I think maybe the easier (if less ideal) thing to do is make ``results.yml`` an include file from ``main.yml``.  Alternatively, if you want to direct me toward the job definition, I can attempt the change myself 😀

cevich commented:
@mrunalp FYI- the "Red X of failure" some jobs may be showing, probably has to do with my comment above to Steve.  This doesn't really affect running the tests at all, just collecting the results.

stevekuznetsov commented:
Right now the results are called in a common post-step for all jobs. Can we write the playbook instead to do as much as possible?

Also, some comments I had about the `results.yml` playbook from the other PR:

 - the playbook should expect to run on the same host as the tests did, not on the Jenkins infrastructure
 - the playbook should not be moving files around or renaming/duplicating them

The framework around your test is logging your output and gathering artifacts that you let us know where you are leaving them. We can handle the fact that Gubernator needs `junit_xx.yml` naming, *etc*, not something the job needs to deal with.

cevich commented:
> Can we write the playbook instead to do as much as possible?

Meaning ``main.yml``?  Yes, I can re-write the results collection to happen there instead.

> the playbook should expect to run on the same host as the tests did, not on the Jenkins infrastructure

There's no "control" host, these all run locally?  Crap, totally missed that one.  Okay, jeeze, right, well I guess it's got to run both ways then, otherwise we build-in that expectation (which is very un-ansible like).  Okay, I'll test that out, thanks for the heads up.

> the playbook should not be moving files around or renaming/duplicating them

"No Moving", that's fine.  Please explain duplicating further, is that because the s3 artifacts loose all sub-directories so the filenames would clash?

@stevekuznetsov Thanks for the advice, I'll re-tool as I encounter the above issues while manual testing.

stevekuznetsov commented:
No, I think your concept of a follow-up playbook that is run as a post-job task is correct. I just meant that this playbook (`results.yml`) could do as much work as it can determine it needs to.

> There's no "control" host, these all run locally? 

There is a control host, but the domain of the test is on the target VM only.

>  Please explain duplicating further, is that because the s3 artifacts loose all sub-directories so the filenames would clash?

If the artifact already exists in the directory we are going to upload, having e.g. the jUnit there twice with different file-names leads to what we're seeing right now where they don't clash and both get uploaded, and we see results x2.

cevich commented:
@stevekuznetsov until I get this PR shaped up, can you kill running that post-task ``results.yml`` playbook?  I think that's an easy short-term fix (so the build-status matches test results).  Let me know if not, and I'll get my arse moving quickly on another PR to fix it.  Thanks again.

cevich commented:
> No, I think your concept of a follow-up playbook that is run as a post-job task is correct.

Oh okay, I can roll with that too (after fixing it)

> There is a control host, but the domain of the test is on the target VM only.

Excellent, I'm glad to hear that.  Okay that should make  #972 easier to swallow.

> If the artifact already exists in the directory we are going to upload, having e.g. the jUnit there twice with different file-names leads to what we're seeing right now where they don't clash and both get uploaded, and we see results x2.

Thanks, that makes more sense and I understand the breakage better.  Okay, I'll re-tool the playbook here (or in another PR) so it's correct.  Sorry about the miscommunication, it seems I was making a lot of assumptions, so it's certainly my bad.

stevekuznetsov commented:
Yeah, I was totally misreading the `results.yml` when I reviewed it as well, sorry! I will now remove the call to `results.yml`.

cevich commented:
Thanks Steve, good thing is we found the problem quickly and the fix(es) seems to be understood all around.

cevich commented:
Rebased from master.  Changed behavior of ``results.py`` playbook:

* Doesn't rely on / require use of any tags, doesn't assume it's running on any particular host.
* Distinguishes between collecting results on the subject and copying them back to the control-host
* When the control-host is the same as the subject, files are only copied around, into ``{{ artifacts }}``.
* When the control-host is different from the subject, additional behaviors occur:

    * A local directory ``{{ collection_dirpath }}/{{ inventory_hostname }}/`` is created[*]
    * The contents from ``{{ artifacts }}`` on each subject are copied there.
    * The ``results.py`` script is run over specific files, producing a single ``{{ collection_dirpath }}/canonical_results.xml`` output file, safe for Jenkins xUnit consumption.

[*] I guess this needs to be changed so the ``{{ inventory_hostname }}`` sub-directory is only used if ``groups.all > 1``?  (i.e. PB is running against an inventory of more than one subject).

cevich commented:
okay, I rebased and ``--autosquash`` from master again and ``--force`` pushed after some manual testing.  Everything is working, ``testout.txt`` from the integration tests is being processed, and combined with ``junit_01.xml`` from the e2e tests.  I sent an example of the output to @stevekuznetsov for checking **before** running the CI tests.

There are some necessary followup changes needed in the jobs themselves (like using the wrapper-script), but I think I may be able to manage that on my own.  If you want to see this working locally it's pretty easy:

0. Stick some files named ``testout.txt`` and/or ``junit_01.xml`` into ``/tmp/artifacts/``
1. ``cd contrib/test/``
2. ``export WORKSPACE=$PWD``
3. ``./venv-ansible-playbook.sh -i localhost, -l localhost -c local --verbose integration/results.yml``
4. Examine ``/tmp/artifacts/junit_01.xml``
5. Eat cake

cevich commented:
Plan B (if steve isn't happy) is to change the output filename (one-line fix) to something other than ``junit_01.xml``, and convince jenkins to consume it but gubinator/grid to not (should be possible).

cevich commented:
Okay, steve says "okay" to combine them.  Let's see whatsheedoo...

/test all

cevich commented:
Rebased onto #972 (which was rebased onto master).  Added a fudging commit for testing purposes.

cevich commented:
(rebase from master & ``--force`` push)

cevich commented:
/test ami

cevich commented:
/test ami

cevich commented:
@mrunalp @rhatdan check it out: Integration and e2e results all in one report!

(click on the 'Show 630 Passed Tests' link)
https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/kubernetes-incubator_cri-o/939/test_pull_request_crio_ami_fedora/69/

/me shakes his fist at travis

cevich commented:
> DO NOT MERGE: FUDGING JENKINS CHANGES HERE FOR TESTING 

Note: I'm working with Steve on the job-changes needed to remove the above commit.

cevich commented:
rebase from #972 & ``--force`` push

cevich commented:
* rebase from #972 & ``--force`` push.
* Includes some more robust handling of jUnit input files under various conditions.
* When input and output are the same file, make a backup before modifying.
* Tweaked the way the subject's hostname is included into the results.
* Tweaked the way additional ID information is included (PR number, commit, etc).

cevich commented:
**N/B:** I'm also testing this manually using an Ansible inventory file (``/tmp/hosts``) containing IPs for a RHEL 7.4 and Fedora 26 VM using:

```
$ cd ./contrib/test
$ export WORKSPACE=$PWD
$ ./venv-ansible-playbook.sh -vv --tags all -i /tmp/hosts integration/main.yml
$ ./venv-ansible-playbook.sh -vv -i /tmp/hosts integration/results.yml
```

Since this is the whole point of this PR and #972 - being able to run properly from an inventory of remote hosts.

cevich commented:
rebase and ``--force`` push.

cevich commented:
rebase from master & ``--force`` push

rhatdan commented:
/test ami

cevich commented:
/test ami

cevich commented:
Note to me: Broke b/c what https://github.com/kubernetes-incubator/cri-o/pull/1119 fixes?

cevich commented:
note to me: testing again w/ 1119 fix (assuming it works)

cevich commented:
/test ami

cevich commented:
Moved into #957
